From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LexManos <LexManos@gmail.com>
Date: Mon, 15 Jun 2020 22:07:51 -0700
Subject: [PATCH] Fix local variables incorrectly merging.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/StackVarsProcessor.java b/src/org/jetbrains/java/decompiler/modules/decompiler/StackVarsProcessor.java
index f62f2f6cba3a5a069d47c414c15fc7aba1202726..4aaf61bd4d2df81bdd1a3f7c314ca030cb62882a 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/StackVarsProcessor.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/StackVarsProcessor.java
@@ -631,7 +631,7 @@ public class StackVarsProcessor {
 
     for (Entry<Integer, Set<VarVersionPair>> ent : mapVars.entrySet()) {
       FastSparseSet<Integer> liveverset = mapLiveVars.get(ent.getKey());
-      if (liveverset == null) {
+      if (liveverset == null || liveverset.isEmpty()) {
         return false;
       }
 
@@ -640,13 +640,13 @@ public class StackVarsProcessor {
         domset.add(ssuversions.nodes.getWithKey(verpaar));
       }
 
-      boolean isdom = false;
+      boolean isdom = true;
 
       for (Integer livever : liveverset) {
         VarVersionNode node = ssuversions.nodes.getWithKey(new VarVersionPair(ent.getKey().intValue(), livever.intValue()));
 
-        if (ssuversions.isDominatorSet(node, domset)) {
-          isdom = true;
+        if (!ssuversions.isDominatorSet(node, domset)) {
+          isdom = false;
           break;
         }
       }
